src/common/w25x.cpp

```diff
@@ -45,26 +50,163 @@ typedef enum {
+ * Try to acquire erase_mutex
+ * acquire communication_mutex and check is_erasing.
+ * If is_erasing pause erase operation
+ * do read/write operation
+ * If is_erasing resume erase operation
+ * communication_mutex must be released first
+ * erase_mutex (if acquired) must be released after communication_mutex is released
+ *
+ * Read operation high priority not implemented / not needed yet.
+ *@{
+ */
+osMutexDef(erase_mutex_resource);
+osMutexId erase_mutex = NULL;
+
+osMutexDef(communication_mutex_resource);
+osMutexId communication_mutex = NULL;
+
+/**@}*/
+
+/**
+ * @brief Block erase operation is in progress
+ *
+ * Block erase operation needs to be paused if read or write is needed immediately.
+ *
+ * This variable can be modified only when both erase_mutex and communication_mutex
+ * are acquired.
+ *
+ * This variable can be read only when communication_mutex is acquired.
+ *
+ * It needs to be checked only when erase_mutex was not acquired successfully, otherwise
+ * holding erase_mutex is sufficient to be sure erasing is not in progress.
+ *
+ * This variable exist because both erase_mutex and communication_mutex must be held
+ * to start erase operation. So erase implementation sets this to true once it acquires both
+ * mutexes. This ensures that high priority read or write operation doesn't assume erase
+ * is already ongoing if it fails to acquire erase_mutex and acquires communication_mutex
+ * in the moment erase acquired erase_mutex and didn't acquired communication_mutex yet for
+ * the first time to start erase operation.
+ *
+ */
+bool is_erasing = false;
+
+class OptionalMutex {
+public:
+    bool acquired() { return m_acquired; }
+
+    OptionalMutex(osMutexId mutex_id, bool wait = true)
+        : m_mutex_id(mutex_id) {
+        if (mutex_id) {
+            auto result = osMutexWait(m_mutex_id, wait ? osWaitForever : 0);
+            m_acquired = result == osOK;
+            if (wait && !m_acquired)
+                bsod("osMutexWait forever failed.");
+        } else {
+            m_acquired = true;
+        }
+    }
+
+    ~OptionalMutex() {
+        if (m_mutex_id && m_acquired) {
+            if (osOK != osMutexRelease(m_mutex_id))
+                bsod("osMutexRelease failed.");
+        }
+    }
+
+private:
+    const osMutexId m_mutex_id;
+    bool m_acquired;
+};
+} //end anonymous namespace
 
 static int w25x_mfrid_devid(uint8_t *devid);
-static void w25x_wait_busy(void);
+static bool w25x_wait_busy(void);
+static bool is_suspended();
+static void resume_erase();
 
 static uint8_t device_id;
 
 bool w25x_init() {
 
-    const bool init_dma = xTaskGetSchedulerState() == taskSCHEDULER_RUNNING;
+    const bool os_running = xTaskGetSchedulerState() == taskSCHEDULER_RUNNING;
+
+    erase_mutex = os_running ? osMutexCreate(osMutex(erase_mutex_resource)) : NULL;
+    communication_mutex = os_running ? osMutexCreate(osMutex(communication_mutex_resource)) : NULL;
```

Comment:
Documentation states that if os is running it can be called exactly once.
> It can be called only once with
> running scheduler as this creates resources which are never released.

Reply:
>  What happens if it is called second time and the mutexes are already initialized?
Init sets pointers to mutexes (pointer is called Id by cmsis to make it more confusing) to NULL. OptionalMutex in case of NULL pointer does nothing - there is no locking.

There is only one use case to call Init without OS running and it is HardFault which is doing core dump to xFlash.

This is needed as HardFault may happen before the OS is started.
Or if the HardFault happens during the time OS is running OS is stopped and xFlash is reinitialized before doing dump.

Reply:
> Furthermore, what happens if they are being held at that time?
In order to be able to call Init if they are held at that time you need to stop the OS first. If you stopped the OS there is no expectation that any code waiting for mutex or holding the mutex will ever run again.
Initialization takes care of stopping any ongoing DMA transfer, ending ongoing operation from the w25x chip point of view. Resuming any paused operation, waiting for finishing that operation and returning at the point w25x is ready to accept any operation (which is expected to be used for crash dump only).

It is good you asked this because while explaining this I noticed one `wait_busy()` call missing there.

Reply:
One more `w25x_wait_busy()` added to `w25x_init()`

Reply:
@dragomirecky Do you have idea if we start the firmware with watchdog enabled? As `w25x_init()` worst case runtime is 200 seconds. If the watchdog is enabled we would get watchdog reset earlier than the return from `w25x_init()`. Maybe we need to add watchdog refresh into `w25x_init()`.